crates/core/tedge_agent/src/file_transfer_server/entity_store.rs

```diff
@@ -0,0 +1,373 @@
+//! This module defines the axum routes and handlers for the entity store REST APIs.
+//! The following endpoints are currently supported:
+//!
+//! - `POST /v1/entities/*path`: Registers a new entity.
+//! - `GET /v1/entities/*path`: Retrieves an existing entity.
+//! - `DELETE /v1/entities/*path`: Deregisters an existing entity.
+//!
+//! References:
+//!
+//! - https://github.com/thin-edge/thin-edge.io/blob/main/design/decisions/0005-entity-registration-api.md
+use super::http_rest::AgentState;
+use crate::entity_manager::server::EntityStoreRequest;
+use crate::entity_manager::server::EntityStoreResponse;
+use axum::extract::Path;
+use axum::extract::State;
+use axum::response::IntoResponse;
+use axum::response::Response;
+use axum::routing::post;
+use axum::Json;
+use axum::Router;
+use hyper::StatusCode;
+use std::str::FromStr;
+use tedge_api::entity::EntityMetadata;
+use tedge_api::entity_store;
+use tedge_api::entity_store::EntityRegistrationMessage;
+use tedge_api::mqtt_topics::EntityTopicId;
+use tedge_api::mqtt_topics::TopicIdError;
+
+#[derive(thiserror::Error, Debug)]
+enum Error {
+    #[error(transparent)]
+    InvalidEntityTopicId(#[from] TopicIdError),
+
+    #[allow(clippy::enum_variant_names)]
+    #[error(transparent)]
+    EntityStoreError(#[from] entity_store::Error),
+
+    #[error("Entity not found with topic id: {0}")]
+    EntityNotFound(EntityTopicId),
+
+    #[allow(clippy::enum_variant_names)]
+    #[error("Failed to publish entity registration message via MQTT")]
+    ChannelError(#[from] tedge_actors::ChannelError),
+
+    #[error("Received unexpected response from entity store")]
+    InvalidEntityStoreResponse,
+}
+
+impl IntoResponse for Error {
+    fn into_response(self) -> Response {
+        let status_code = match &self {
+            Error::InvalidEntityTopicId(_) => StatusCode::BAD_REQUEST,
+            Error::EntityStoreError(_) => StatusCode::BAD_REQUEST,
+            Error::EntityNotFound(_) => StatusCode::NOT_FOUND,
+            Error::ChannelError(_) => StatusCode::INTERNAL_SERVER_ERROR,
+            Error::InvalidEntityStoreResponse => StatusCode::INTERNAL_SERVER_ERROR,
+        };
+        let error_message = self.to_string();
+
+        (status_code, error_message).into_response()
+    }
+}
+
+pub(crate) fn entity_store_router(state: AgentState) -> Router {
+    Router::new()
+        .route(
+            "/v1/entities/*path",
+            post(register_entity)
```

Comment:
suggestion: Have `POST` for `v1/entities`, not `v1/entities/*path`

In the proposal, `POST` is declared for `v1/entities` collection, not the resource, so is the change here desired or a leftover from a simpler implementation?
Putting this route on a collection level would also fix the "having to put topic id twice" problem.
If there is a good reason for requiring a path here, it should be documented.

Reply:
I've already fixed this already, following Didier's comments. It came from my desire to keep the topics of all the CRUD APIs consistent during the early days. But, following a discussion with the team, we had agreed on dropping the topic id from the path for this POST request.  It just wasn't cleaned up from the impl.